In [3]:
from math import sqrt
from scipy import stats

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from pydataset import data
import statistics
import acquire
import prepare

### Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [12]:
df = pd.read_csv("untidy-data/attendance.csv")

In [13]:
df

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [14]:
df = df.melt(id_vars=['Unnamed: 0'], var_name='date', value_name='value')

In [26]:
df.head(3)

Unnamed: 0        date value
0      Sally  2018-01-01     P
1       Jane  2018-01-01     A
2      Billy  2018-01-01     A

In [49]:
df.value.unique()

array(['P', 'A', 'T', 'H'], dtype=object)

In [59]:
def score(value):
    if value == "P":
        return 1
    elif  value == "A":
        return 0
    elif  value == "T":
        return .90
    elif  value == "H":
        return .5
        

In [60]:
df["score"] = df.value.apply(score)

In [61]:
df.head(3)

Unnamed: 0        date value  score
0      Sally  2018-01-01     P    1.0
1       Jane  2018-01-01     A    0.0
2      Billy  2018-01-01     A    0.0

In [65]:
df = df.rename(columns={'Unnamed: 0': 'name'})

In [66]:
df.groupby("name").score.mean()

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: score, dtype: float64

Read the coffee_levels.csv file.

In [78]:
df = pd.read_csv("untidy-data/coffee_levels.csv")

In [79]:
df.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [80]:
df.coffee_carafe.unique()

array(['x', 'y', 'z'], dtype=object)

Transform the data so that each carafe is in it's own column.

In [81]:
df = df.pivot_table(index = 'hour',
                    columns = 'coffee_carafe',
                    values = 'coffee_amount').reset_index()

In [83]:
df

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

Is this the best shape for the data?

Probably the orginal shape it came in.  The transposed is a bit tougher to make sense of. 

### Read the cake_recipes.csv data.
### This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [131]:
df = pd.read_csv("untidy-data/cake_recipes.csv")

Tidy the data as necessary.

In [132]:
df.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [133]:
df = df.melt(id_vars=['recipe:position'], var_name='date', value_name='value')

In [121]:
df.head(3)

recipe:position date      value
0        a:bottom  225  61.738655
1           a:top  225  51.709751
2        b:bottom  225  57.095320

In [134]:
df[['recipe', 'position']]  = df["recipe:position"].str.split(':', expand = True)

In [123]:
df.head(3)

recipe:position date      value recipe position
0        a:bottom  225  61.738655      a   bottom
1           a:top  225  51.709751      a      top
2        b:bottom  225  57.095320      b   bottom

In [135]:
df = df.drop(columns='recipe:position')

In [136]:
df.head()

date      value recipe position
0  225  61.738655      a   bottom
1  225  51.709751      a      top
2  225  57.095320      b   bottom
3  225  82.455004      b      top
4  225  96.470207      c   bottom

In [175]:
df_r_p = df.pivot_table(index = 'date',
                    columns = ['recipe', 'position'],
                    values = 'value').reset_index()

In [176]:
df_r_p

recipe   date          a                     b                     c  \
position          bottom        top     bottom        top     bottom   
0         225  61.738655  51.709751  57.095320  82.455004  96.470207   
1         250  53.912627  52.009735  61.904369  95.224151  52.001358   
2         275  74.414730  68.576858  61.196980  98.594881  92.893227   
3         300  98.786784  50.228470  99.248541  58.169349  65.473084   

recipe                       d             
position        top     bottom        top  
0         71.306308  52.799753  96.873178  
1         82.795477  58.670419  76.101363  
2         92.098049  51.747686  59.571620  
3         53.960273  56.183110  50.971626

In [138]:
df_r = df.pivot_table(index = 'date',
                    columns = 'recipe',
                    values = 'value').reset_index()

In [140]:
df_r.head()

recipe date          a          b          c          d
0       225  56.724203  69.775162  83.888258  74.836465
1       250  52.961181  78.564260  67.398417  67.385891
2       275  71.495794  79.895930  92.495638  55.659653
3       300  74.507627  78.708945  59.716678  53.577368

In [153]:
df_r.mean()

recipe
date    5.631257e+10
a       6.392220e+01
b       7.673607e+01
c       7.587475e+01
d       6.286484e+01
dtype: float64

Which recipe, on average, is the best? recipe b

In [171]:
df_r[["a", 'b', 'c', 'd']].mean().sort_values(ascending=False)

recipe
b    76.736074
c    75.874748
a    63.922201
d    62.864844
dtype: float64

Which oven temperature, on average, produces the best results? 275

In [208]:
df_r_p

recipe   date          a                     b                     c  \
position          bottom        top     bottom        top     bottom   
0         225  61.738655  51.709751  57.095320  82.455004  96.470207   
1         250  53.912627  52.009735  61.904369  95.224151  52.001358   
2         275  74.414730  68.576858  61.196980  98.594881  92.893227   
3         300  98.786784  50.228470  99.248541  58.169349  65.473084   

recipe                       d             
position        top     bottom        top  
0         71.306308  52.799753  96.873178  
1         82.795477  58.670419  76.101363  
2         92.098049  51.747686  59.571620  
3         53.960273  56.183110  50.971626

In [209]:
df_r_p.iloc[:, 1:].mean(axis=1)

0    71.306022
1    66.577437
2    74.886754
3    66.627655
dtype: float64

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [229]:
df.iloc[df.value.idxmax()]

date              300
value       99.248541
recipe              b
position       bottom
Name: 26, dtype: object

recipe b, bottom rack, 300 degrees